x-pack/plugins/apm/server/lib/service_map/get_service_map_from_trace_ids.test.ts

```diff
@@ -10,19 +10,76 @@ import { PromiseReturnType } from '../../../typings/common';
 import { fetchServicePathsFromTraceIds } from './fetch_service_paths_from_trace_ids';
 
 describe('getConnections', () => {
-  it('transforms a list of paths into a list of connections filtered by service.name and environment', () => {
-    const response = serviceMapFromTraceIdsScriptResponse as PromiseReturnType<
-      typeof fetchServicePathsFromTraceIds
-    >;
-    const serviceName = 'opbeans-node';
-    const environment = 'production';
-
-    const connections = getConnections(
-      response.aggregations?.service_map.value.paths,
-      serviceName,
-      environment
-    );
-
-    expect(connections).toMatchSnapshot();
+  describe('transforms a list of paths into a list of connections', () => {
+    it('with no filters', () => {
+      const response = serviceMapFromTraceIdsScriptResponse as PromiseReturnType<
+        typeof fetchServicePathsFromTraceIds
+      >;
+
+      const connections = getConnections(
+        response.aggregations?.service_map.value.paths
+      );
+
+      expect(connections).toMatchSnapshot();
+    });
+    it('filters by service.name and environment', () => {
+      const response = serviceMapFromTraceIdsScriptResponse as PromiseReturnType<
+        typeof fetchServicePathsFromTraceIds
+      >;
+      const serviceName = 'opbeans-node';
+      const environment = 'production';
+
+      const connections = getConnections(
+        response.aggregations?.service_map.value.paths,
+        serviceName,
+        environment
+      );
+
+      expect(connections).toMatchSnapshot();
+    });
+
+    it('filters by service.name', () => {
+      const response = serviceMapFromTraceIdsScriptResponse as PromiseReturnType<
+        typeof fetchServicePathsFromTraceIds
+      >;
+      const serviceName = 'opbeans-node';
+
+      const connections = getConnections(
+        response.aggregations?.service_map.value.paths,
+        serviceName
+      );
+
+      expect(connections).toMatchSnapshot();
+    });
+
+    it('filters by environment', () => {
+      const response = serviceMapFromTraceIdsScriptResponse as PromiseReturnType<
+        typeof fetchServicePathsFromTraceIds
+      >;
+      const environment = 'testing';
+
+      const connections = getConnections(
+        response.aggregations?.service_map.value.paths,
+        undefined,
+        environment
+      );
+
+      expect(connections).toMatchSnapshot();
+    });
+
+    it('filters by environment not defined', () => {
+      const response = serviceMapFromTraceIdsScriptResponse as PromiseReturnType<
+        typeof fetchServicePathsFromTraceIds
+      >;
+      const environment = 'ENVIRONMENT_NOT_DEFINED';
+
+      const connections = getConnections(
+        response.aggregations?.service_map.value.paths,
+        undefined,
+        environment
+      );
+
+      expect(connections).toMatchSnapshot();
```

Comment:
Having a single snapshot test is fine but most of the tests should be more specific, eg

```ts
describe('if environment is "not defined"', () => {
  it('does not include connections with source environment set', () => {
    expect(connections.every((conn) => !conn.source['service.environment'])).toBe(true);
  })
})

```

Reply:
We should specifically verify the nitty gritty implementation details of `getConnections` because it's very easy to modify it without noticing subtle changes.

Reply:
@sqren most of these tests return a lot of data, which is not easy to test specific things. I have changed one that filters by the environment, but the others I'd rather keep on snapshot if that's with you.

Reply:
I think the tests could still be more explicit. eg. "filters by service.name and environment". What is the expected behaviour here? I should be able to understand the implementation by reading the test.

I still think tests along these lines are better:
```ts
describe('if environment is "not defined"', () => {
  it('does not include connections with source environment set', () => {
```

```ts
describe('if environment is "all" (missing)', () => {
  it('includes both connections with and without source environment set', () => {
```

```ts
describe('if service name is given', () => {
  it('excludes connections with a different service name', () => {
```

```ts
describe('if service name is not given', () => {
  it('includes connections with different service names', () => {
```